Vishnu/user dict

vvijay874 opened the pull request:
Created a full user dictionary using a SQLite database
Created a full sign up / login functionality
Created a UI for the socialBot and fileUpload page
Separated the socialBot and fileUpload pages using a BrowserRouter
Created a unique socialBot object for each user and allows them to access their specific object

vvijay874 committed:
Created user dictionary

vvijay874 committed:
Created a simple frontend to create account

Created a signup with a simple username system

vvijay874 committed:
Completely Functional connection between user and socialbot

Created a connection between the social bot class and user dictionary. Now there is a new socialBot object per user, and this allows for unique responses based on each user

vvijay874 committed:
UI created for the social bot page

UI created and functional!

vvijay874 committed:
Created a full login functionality

Logging in fully works, with the text area updating if you log in to another account.

vvijay874 committed:
Merge branch 'main' into vishnu/user-dict

vvijay874 committed:
Created a full routing system

The pages are now separated and can be accessed separately. The file upload page is alone and the socialbot page is alone. They do not depend on one another and are routed using the BrowserRouter

vvijay874 committed:
Created SQLite database for users

Created a fully functional SQLite database for the users that works.

MichaelDeng03 commented on client/src/App.css:
```diff
@@ -13,7 +13,7 @@
 .chat-box {
   flex: 1;
-  overflow-y: auto;
+  /* overflow-y: auto; */
```
delete commented out code

MichaelDeng03 commented on socialBot.py:
```diff
@@ -14,13 +11,20 @@
-user_dict = {
-    
-}
+socialBotArr = []
```
We talked about this during our meeting, but you may be able to serialize your SocialBot class, and init it on the fly.
I wouldn't worry too much about optimization, but do try to prevent possible infinite memory usage situations.

MichaelDeng03 commented on socialBot.py:
```diff
@@ -14,13 +11,20 @@
+command1 = """CREATE TABLE IF NOT EXISTS 
+users(username TEXT PRIMARY KEY, socialind INTEGER)"""
+cursor.execute(command1)
+cursor.execute("SELECT * FROM users")
```
Definitely don't select all the users.

MichaelDeng03 commented on socialBot.py:
```diff
@@ -14,13 +11,20 @@
-}
+socialBotArr = []
+
+connection = sqlite3.connect('user_information.db', check_same_thread=False)
```
I'd like to see the SQL management happen in another module.
In real-world environments, your server won't be making the DB, and the DB will be stood up and updaterd by another process (probably another person entirely), and your socialBot will have limited permissions.
That being said, this can be a giant sinkhole of your time.
So, I think you should do the minimum, which is to separate the table creation into a separate script (so just copy and paste this code), and drop a comment/README telling users to run the script before starting the server.

MichaelDeng03 commented on socialBot.py:
```diff
@@ -101,16 +105,37 @@ def sendInfo(self):
+    social_bot = None
+    for x in rows:
+        if x[0] == currentUser:
+            social_bot = socialBotArr[x[1]]
```
really innovative solution, but the better solution would be to init the socialbot as needed.
something like a many (socialbot) to one (user) relationship would be apt.

MichaelDeng03 commented on socialBot.py:
```diff
@@ -101,16 +105,37 @@ def sendInfo(self):
     message = request.json.get('input_data')   
-    social_bot = user_dict.get(currentUser)
+    social_bot = None
+    for x in rows:
```
🤮 
the SQL `SELECT` solution is the clearest 'good' solution.
The python only equivalent would be to use a dictionary (applies for your socialBotArr as well)
Don't do this, but this may help you visualize the potential relationships
```
members: dict[str, dict[str, SocialBot]] = {}
social_bots: dict[str, SocialBot] = members[currentUser]
social_bot: SocialBot = social_bots[social_bot_username]
```

MichaelDeng03 commented on socialBot.py:
```diff
@@ -101,16 +105,37 @@ def sendInfo(self):
+        if x[0] == potUser:
+            return jsonify({"result": "That user already exists! Try again!"})
+    currentUser = potUser
+    cursor.execute(f"INSERT INTO users VALUES ('{currentUser}', {len(socialBotArr)})")
```
Throw this insert statement into a function which accepts a cursor and a currentUser.

MichaelDeng03 commented on socialBot.py:
```diff
@@ -14,13 +11,20 @@
+connection = sqlite3.connect('user_information.db', check_same_thread=False)
+cursor = connection.cursor()
+command1 = """CREATE TABLE IF NOT EXISTS 
+users(username TEXT PRIMARY KEY, socialind INTEGER)"""
```
It wasn't clear to me what socialind was before reading the rest of the code. 
I think you'll end up removing this column, and replacing it with a SocialBot table.
I won't give you the answer, but this table will need a foreign key for creating a many (socialbot) to one relationship (user)

MichaelDeng03 commented on user_information.db:
Don't commit db files, or db-journal files.
 
I think you're good to do `git rm --cached <file_name>`, and add it to your gitignore. 

Note, however, that this will leave the db in your git history. Meaning, only your commits after this removal will show the db as not existing, but past commits will still have it. 

This isn't a huge issue because we `squash and merge`, so only the end state of this branch gets merged into main, and not the intermediary commits. If we didn't squash and merge, this db would forever appear in the commit history of main.

MichaelDeng03 commented:
In general, try to keep the scope of the PR tighter.
I appreciate the effort that went into the refactor/making the pages look cleaner, but if your goal was to create a user-socialbot association, try to keep that the main focus.

vvijay874 committed:
SQLite database changed

Shifted away from the array of socialBot objects and instead store information to make a new SocialBot object from the SQLite database itself

SVPandya commented on socialBot.py:
```diff
@@ -98,19 +108,48 @@ def sendInfo(self):
+    return jsonify({"result": "Login unsuccessful! No user found!"})
 
 if __name__ == "__main__":
     app.run(debug=True)
```
Overall comment: split the SQL database separately from socialBot.py so it can maybe be used in other files.
